src/node/tendermint_processes.ml

```diff
@@ -0,0 +1,367 @@
+(** Tendermint's consensus processes as describded at https://arxiv.org/pdf/1807.04938.pdf *)
+
+open Tendermint_internals
+open Tendermint_data
+
+module type Clock_type = sig
+  type 'a t = {
+    time : int;
+    step : consensus_step;
+    height : height;
+    round : round;
+    on_timeout : consensus_state -> input_log -> OutputLog.t -> State.t -> 'a;
+    started : bool;
+  }
+  val make :
+    int ->
+    consensus_step ->
+    height ->
+    round ->
+    (consensus_state -> input_log -> OutputLog.t -> State.t -> 'a) ->
+    'a t
+end
+
+module Clock : Clock_type = struct
+  type 'a t = {
+    time : int;
+    step : consensus_step;
+    height : height;
+    round : round;
+    on_timeout : consensus_state -> input_log -> OutputLog.t -> State.t -> 'a;
+    started : bool;
+  }
+
+  let make time step height round on_timeout =
+    { time; step; on_timeout; started = false; height; round }
+end
+
+type consensus_action =
+  | Broadcast         of sidechain_consensus_op
+  | Schedule          of consensus_action option Clock.t
+  | RestartTendermint of (height * round)
+  | DoNothing
+
+type process =
+  height ->
+  round ->
+  consensus_state ->
+  input_log ->
+  OutputLog.t ->
+  State.t ->
+  consensus_action option
+
+type clock = consensus_action option Clock.t
+
+let on_timeout_propose (height : height) (round : round)
+    (consensus_state : consensus_state) (msg_log : input_log) _dlog
+    _global_state =
+  let do_something consensus_state =
+    if
+      height = consensus_state.height
+      && round = consensus_state.round
+      && consensus_state.step = Proposal
+    then (
+      consensus_state.step <- Prevote;
+      Broadcast (PrevoteOP (consensus_state.height, consensus_state.round, nil)))
+    else
+      DoNothing in
+  if contains_timeout msg_log height Proposal then
+    Some (do_something consensus_state)
+  else
+    None
+
+let on_timeout_prevote (height : height) (round : round)
+    (consensus_state : consensus_state) (msg_log : input_log) _dlog
+    _global_state =
+  let do_something consensus_state =
+    if
+      height = consensus_state.height
+      && round = consensus_state.round
+      && consensus_state.step = Prevote
+    then (
+      consensus_state.step <- Precommit;
+      Broadcast
+        (PrecommitOP (consensus_state.height, consensus_state.round, nil)))
+    else
+      DoNothing in
+  if contains_timeout msg_log height Prevote then
+    Some (do_something consensus_state)
+  else
+    None
+
+let start_round (height : height) (round : round)
+    (consensus_state : consensus_state) (_msg_log : input_log) _dlog
```

Comment:
_msg_log and _dlog?

Reply:
Unused here, however, we need all the Tendermint process functions to have the same signature.